src/System.Management.Automation/engine/parser/TypeResolver.cs

```diff
@@ -330,10 +369,23 @@ internal static Type ResolveTypeNameWithContext(TypeName typeName, out Exception
                         assemList = ClrFacade.GetAssemblies(typeResolutionState, newTypeName);
                     }
 #endif
-                    var newResult = ResolveTypeNameWorker(newTypeName, currentScope, typeResolutionState.assemblies, typeResolutionState, /* reportAmbiguousException */ true, out exception);
+                    var newResult = ResolveTypeNameWorker(newTypeName, currentScope, typeResolutionState.assemblies, searchedAssemblies, typeResolutionState,
+                                                          /*onlySearchInGivenAssemblies*/ false, /* reportAmbiguousException */ true, out exception);
                     if (exception == null && newResult == null)
                     {
-                        newResult = ResolveTypeNameWorker(newTypeName, currentScope, assemList, typeResolutionState, /* reportAmbiguousException */ false, out exception);
+                        if (context != null && assemblies == null)
+                        {
+                            // If the assemblies to search from is not specified by the caller, then we search our assembly cache first,
+                            // so as to give preference to resolving the type against assemblies explicitly loaded by powershell.
+                            newResult = ResolveTypeNameWorker(newTypeName, currentScope, context.AssemblyCache.Values, searchedAssemblies, typeResolutionState,
```

Comment:
This looks like the same search pattern used above.  Can this be made into a helper method and reduce code?